### PR TITLE
Cce analytic test abstract base class

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Introduces the abstract base class to provide analytic worldtube data and the expected news for analytic data tests of CCE.

edit: The bouncing black hole part has been split out for a different PR.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/Systems/Cce/AnalyticSolutions/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  )
+
+add_spectre_library(${LIBRARY} ${LIBRARY_SOURCES})
+
+target_link_libraries(
+  ${LIBRARY}
+  PUBLIC
+  DataStructures
+  ErrorHandling
+  IO
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/CMakeLists.txt:19]: GeneralizedHarmonic and GeneralRelativity because of the tags

File: src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+                  gr::Tags::SpatialMetric<3, ::Frame::Inertial, DataVector>>,
+              ::Tags::dt<gr::Tags::Shift<3, ::Frame::Inertial, DataVector>>,
+              ::Tags::dt<gr::Tags::Lapse<DataVector>>,
+              Tags::Dr<
+                  gr::Tags::SpatialMetric<3, ::Frame::Inertial, DataVector>>,
+              Tags::Dr<gr::Tags::Shift<3, ::Frame::Inertial, DataVector>>,
+              Tags::Dr<gr::Tags::Lapse<DataVector>>, Tags::News>,
+          tmpl::bind<IntermediateCacheTag, tmpl::_1>>>;
+
+  mutable IntermediateCacheTuple intermediate_cache_;
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp:254]: we are not happy this has internal state; isn't there more than one component? our other analytic solutions intentionally avoid having internal state

File: src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        Spectral::Swsh::number_of_swsh_collocation_points(output_l_max));
+    variables_impl(make_not_null(&item), output_l_max, time, Tag{});
+    item_cache.l_max = output_l_max;
+    item_cache.time = time;
+    return item;
+  }
+  virtual void prepare_solution(size_t output_l_max, double time) const
+      noexcept = 0;
+
+  // note that template functions cannot be virtual, so we have to emulate
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp]: function templates

File: src/Evolution/Systems/Cce/Tags.hpp [hunk 4]
```diff
@@ -124,7 +124,7 @@ struct Dy : db::PrefixTag, db::SimpleTag {
 /// The derivative with respect to Bondi \f$r\f$
 template <typename Tag>
 struct Dr : db::PrefixTag, db::SimpleTag {
-  using type = Scalar<SpinWeighted<ComplexDataVector, Tag::type::type::spin>>;
+  using type = db::item_type<Tag>;
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Tags.hpp]: why this change?

**Review by kidder — CHANGES_REQUESTED** [review]
reviewed with @nilsdeppe 
not a thorough review as the caching scares us because it is not thread-safe.

File: src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp (hunk 2 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp:254]: Some of the caching could be omitted in favor of recomputing many of the inputs several times depending on the combination requested -- that would make this class comparatively slow, but have similar interface.
I suppose I could have each of the variable generation steps also take an argument where the calling code holds the cache and passes it in to improve the speed, but then the interface becomes more cumbersome by requiring the calling code manage the data storage etc.
This is particularly relevant for a later "analytic" solution that I'll be putting in later that follows this framework -- the Robinson-Trautman solution that actually requires an internal evolved variable to determine the solution. My view is that version really should have internal state, because otherwise the calling code has to perform far too much management of the time-stepped variable.
Can you clarify the question about more than one component? This class is intended to only be held by a different version of the `WorldtubeBoundary` (singleton) component, so the internal state data would not be duplicated across components.

File: src/Evolution/Systems/Cce/Tags.hpp (hunk 4 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/Tags.hpp]: In the new `WorldtubeData` class, I represent the (Bondi-Sachs) radial derivative of tensor quantities like the spatial metric, so it doesn't work to assume that it prefixes a spin-weighted quantity anymore.

**Comment by moxcodes** [comment]
I've added a commit that fixes some of the easy changes and hopefully clarifies the role of the analytic data in the CCE system.
Thanks for the review @kidder and @nilsdeppe !

File: src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+   * tags is found in `WorldtubeData::tags`, and includes coordinate and
+   * Jacobian quantities as well as metric quantities and derivatives thereof.
+   */
+  template <typename... Tags>
+  tuples::TaggedTuple<Tags...> variables(const size_t output_l_max,
+                                         const double time,
+                                         tmpl::list<Tags...> /*meta*/) const
+      noexcept {
+    prepare_solution(output_l_max, time);
+    return {cache_or_compute(output_l_max, time, Tags{})...};
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp]: `cache_or_compute<Tags>(output_l_max, time)...`

File: src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+  }
+
+ protected:
+  template <typename Tag>
+  db::const_item_type<Tag>& cache_or_compute(size_t output_l_max, double time,
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp]: const auto& as return type   
const the arguments.

File: src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+                                             Tag /*meta*/) const noexcept {
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp]: remove this argument

File: src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    auto& item_cache = get<IntermediateCacheTag<Tag>>(intermediate_cache_);
+    if (item_cache.l_max == output_l_max and item_cache.time == time) {
+      return item_cache.data;
+    }
+    auto& item = item_cache.data;
+    destructive_resize_components(
+        make_not_null(&item),
+        Spectral::Swsh::number_of_swsh_collocation_points(output_l_max));
+    variables_impl(make_not_null(&item), output_l_max, time, Tag{});
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp]: Tag{} should be either `tmpl::list<Tag>` or `tmpl::type_<Tag>` to future-proof against tags having state

File: src/Evolution/Systems/Cce/Tags.hpp (hunk 4 above)
Comment (kidder) [reply src/Evolution/Systems/Cce/Tags.hpp]: just make it `typename Tag::type`

File: src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.cpp [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace Cce {
+namespace Solutions {
+
+/// \cond
+void WorldtubeData::variables_impl(
+    gsl::not_null<tnsr::i<DataVector, 3>*> cartesian_coordinates,
+    size_t output_l_max, double /*time*/,
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.cpp]: const everything that should be const everywhere 😛

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by moxcodes** [comment]
fixups posted.
Cheers!

**Review by kidder — COMMENTED** [review]
I am happy with the fixups... @nilsdeppe ?

**Comment by nilsdeppe** [comment]
yep, looks good. Please squash

**Comment by moxcodes** [comment]
Squashed.
Cheers!

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
macOS failed to build, I don't understand why. CI just says `The operation was cancelled`
